src/Voron/Util/ContextBoundNativeList.cs

```diff
@@ -0,0 +1,123 @@
+﻿using System;
+using Sparrow.Server;
+
+namespace Voron.Util
+{
+    public unsafe struct ContextBoundNativeList<T> : IDisposable
+        where T : unmanaged
+    {
+        private readonly ByteStringContext _ctx;
+        public NativeList<T> Inner;
```

Comment:
No, it is very error prone. Added more information about why `CountextBoundNativeList<T>` is needed and the preferred usage for it. `NativeList<T>` is too low level for normal usage.

Reply:
Why would that be error prune? My point is that we can avoid mostly duplicated work.
For the case of the `NativeList<NativeList<RecordedTerm>>` - there is no reason why we have to accept this.
Just turn that into a `NativeList<RecordedTerm>[]` and be done with it. There is not going to be a lot of traffic there and we can reuse the buffer easily enough across invocation.

Reply:
What about `NativeList<>[]` and having a private ArrayPool for it?  GC should be less expensive in such case.

Reply:
That is what I would rather do, yes.
In practice, the issue is that this list may be _very_ big (per term).
So we might actually want to be wary about killing it afterward or sharing it across all indexes or putting a size limit on the total array size to avoid keeping stuff in memory forever in the case of a single big batch

Reply:
Killing it afterwards now is as simple as calling `.Reset()` on the `ByteStringContext` ... that's why I dont like the alternative of having to deal with managed objects.